### PR TITLE
Automatically convert MySQL/MariaDB application DBs to utf8mb4 on launch

### PR DESCRIPTION
**Pull request by camsaul**
Extra fix related to #10691 

This is a function that actually converts an entire MySQL database to the `utf8mb4` encoding (if needed). It's working and tested, and optimized a bit to skip tables that don't need to be converted. 

I'm not sure if the performance of this will be reasonable or not for very large datasets, I need to experiment with that and confirm.

### CODE REVIEW

File: src/metabase/db.clj [hunk 1]
```diff
@@ -262,8 +262,8 @@
 
   ([driver :- s/Keyword, details :- su/Map]
    (log/info (u/format-color 'cyan (trs "Verifying {0} Database Connection ..." (name driver))))
+   (classloader/require 'metabase.driver.util)
```
Comment (walterl) [inline src/metabase/db.clj]: Why was this moved out of the `assert`?

File: src/metabase/db/fix_mysql_utf8.clj [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                   {:character-set "utf8mb4", :collation "utf8mb4_unicode_ci"})
+              (log/debug "Default character set for application database is already in utf8mb4")
+              [(format "ALTER DATABASE `%s` CHARACTER SET = utf8mb4 COLLATE = utf8mb4_unicode_ci;" database-name)])
+            (jdbc/with-db-metadata [metadata conn]
+              (->> (table-names metadata database-name)
+                   (map (fn [table-name]
+                          (if (= (table-character-set-and-collation conn database-name table-name)
+                                 {:character-set "utf8mb4", :collation "utf8mb4_unicode_ci"})
+                            (log/debug "Character set for table '%s' is already utf8mb4" table-name)
+                            (format "ALTER TABLE `%s` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;"
```
Comment (walterl) [inline src/metabase/db/fix_mysql_utf8.clj]: Don't we also need to convert specific [columns](https://dev.mysql.com/doc/refman/5.7/en/charset-column.html) that are using different character sets?

File: src/metabase/db/fix_mysql_utf8.clj [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                                    table-name))))
+                   (filter some?))))))]
+    (when (seq statements)
+      (concat ["SET foreign_key_checks = 0;"]
+              statements
+              ["SET foreign_key_checks = 1;"]))))
```
Comment (walterl) [inline src/metabase/db/fix_mysql_utf8.clj]: If another, previous statement errors out, is there something that guarantees that `foreign_key_checks` will be re-enabled?

For interest's sake, why are you disabling this?

**Review by walterl — COMMENTED** [review]
I haven't read the related issue, but what's in the PR looks good. I just have a few questions.

File: src/metabase/db/fix_mysql_utf8.clj (hunk 2 above)
Comment (camsaul) [reply src/metabase/db/fix_mysql_utf8.clj]: No, you just need to do the tables. Converting the tables converts the columns as well. It's verified in the tests

File: src/metabase/db/fix_mysql_utf8.clj (hunk 3 above)
Comment (camsaul) [reply src/metabase/db/fix_mysql_utf8.clj]: something I read somewhere said you had to disable FK checks when converting the character encodings. I'll try to find out where I read that

Comment (camsaul) [reply src/metabase/db/fix_mysql_utf8.clj]: but you are right, it should be in a `finally` block at the very least

Comment (camsaul) [reply src/metabase/db/fix_mysql_utf8.clj]: This is the reason you'd want to disable FK checks

https://dba.stackexchange.com/questions/154180/converting-character-sets-and-foreign-key-constaints?rq=1

File: src/metabase/db.clj (hunk 1 above)
Comment (camsaul) [reply src/metabase/db.clj]: It doesn't really make a difference actually, if it throws an Exception it will make no difference whether it was in the `assert` or not. I moved it up to make it clearer that the `binding` isn't intended to have an effect on that expression

File: src/metabase/db/fix_mysql_utf8.clj (hunk 3 above)
Comment (camsaul) [reply src/metabase/db/fix_mysql_utf8.clj]: I ended up tweaking this to run the migrations in an order where disabling FK checks is no longer required
